contracts/token/NonFungibleToken.sol

```diff
@@ -0,0 +1,289 @@
+  * @param _tokenId uint256 ID of the token to query the owner of
+  * @return owner address currently marked as the owner of the given token ID
+  */
+  function ownerOf(uint256 _tokenId) public view returns (address owner) {
+    owner = tokenOwner[_tokenId];
+    require(owner != address(0));
+  }
+
+  /**
+  * @dev Gets the token ID at a given index of the tokens list of the requested owner
+  * @param _owner address owning the tokens list to be accessed
+  * @param _index uint256 representing the index to be accessed of the requested tokens list
+  * @return uint256 token ID at the given index of the tokens list owned by the requested address
+  */
+  function tokenOfOwnerByIndex(address _owner, uint256 _index) public view returns (uint256) {
+    require(_index < balanceOf(_owner));
+    return ownedTokens[_owner][_index];
+  }
+
+  /**
+   * @dev Gets the approved address to take ownership of a given token ID
+   * @param _tokenId uint256 ID of the token to query the approval of
+   * @return address currently approved to take ownership of the given token ID
+   */
+  function approvedFor(uint256 _tokenId) public view returns (address) {
+    return tokenApprovals[_tokenId];
+  }
+
+  /**
+   * @dev Tells whether an operator is approved by a given owner
+   * @param _owner owner address which you want to query the approval of
+   * @param _operator operator address which you want to query the approval of
+   * @return bool whether the given operator is approved by the given owner
+   */
+  function isOperatorApprovedFor(address _owner, address _operator) public view returns (bool) {
+    return operatorApprovals[_owner][_operator];
+  }
+
+  /**
+  * @dev Transfers the ownership of a given token ID to another address
+  * @param _to address to receive the ownership of the given token ID
+  * @param _tokenId uint256 ID of the token to be transferred
+  */
+  function transfer(address _to, uint256 _tokenId) public onlyOwnerOf(_tokenId) {
+    clearApprovalAndTransfer(msg.sender, _to, _tokenId);
+  }
+
+  /**
+   * @dev Transfer the ownership of a given token ID to another contract address and calls it with given data
+   * @param _to address to receive the ownership of the given token ID
+   * @param _tokenId uint256 ID of the token to be transferred
+   * @param _data bytes of data to be sent when the receiving address gets called
+   */
+  function transfer(address _to, uint _tokenId, bytes _data) public onlyOwnerOf(_tokenId) {
+    clearApprovalAndTransfer(msg.sender, _to, _tokenId);
+    require(_to.call.value(0)(_data));
+  }
+
+  /**
+  * @dev Approves an address to claim ownership of any token owned by the msg.sender
+  * @param _to operator address to set their approval
+  * @param _approved representing the status of the approval
+  */
+  function approve(address _to, bool _approved) public {
+    require(_to != msg.sender); // we are reverting here following the same behaviour as approve does
```

Comment:
The comment is confusing because this function is called `approve` as well. I'm not comfortable with that... Have we considered other names for this one?